### PR TITLE
Add proxy request/response completion handlers

### LINKED ISSUE #29: Completion handler for user agent responses
Hi All,

Thanks for creating and maintaining a very useful project! I would like to suggest the following feature which we think would be a useful addition (see use cases below).

- [x] do not use this issue tracker to ask questions, instead use one of these [channels](https://vertx.io/community/). Questions will likely be closed without notice.
- [x] you shall create a feature request only when it is general purpose enough.
- [x] make sure that the feature is not already implemented

#### Describe the feature

A completion handler such as:
```
  @Fluent
  HttpProxy onProxyResponseCompletion(Handler<AsyncResult<Void>> handler);
```

(or with an equivalent origin request completion handler):
```
  @Fluent
  HttpProxy onProxyRequestCompletion(Handler<AsyncResult<Void>> handler);

  @Fluent
  HttpProxy onProxyResponseCompletion(Handler<AsyncResult<Void>> handler);
```
which are called when the request to the origin is completed: `onProxyRequestCompletion` and when the response to the user agent is completed: `onProxyResponseCompletion`.

#### Use cases

* Track pending/outstanding proxy requests to allow graceful shutdown of the proxy:
  * Each time `handle(HttpServerRequest outboundRequest)` is called increment a counter in the application code
  * Each time the completion handler is called, decrement the counter in `onProxyResponseCompletion`.
  * Before shutting down the proxy, block new incoming requests, wait until the counter reaches zero, finally shut down the proxy.

* Gather latency metrics on overall request proxy time
  * Before calling `handle` start a timer
  * Stop the timer In the completion handler: `onProxyResponseCompletion`
  * Report the proxy latency as the difference between the timers

#### Contribution

I am volunteering for implementing this feature.

### PR DESCRIPTION
**Pull request by dleen**
Closes #29

Motivation:

Looking to get feedback on the approach as an implementation of https://github.com/eclipse-vertx/vertx-http-proxy/issues/29

Conformance:

I've signed the eclipse agreement and my IDE uses editorconfig.

### CODE REVIEW

**Comment by dleen** [comment]
I think the use of `Future<Void>` is not very helpful here as the request/response we are acting on in the callback has already been resolved when the handlers are called. It is useful for signaling the success/failure however. What do you suggest replacing it with? If a user wanted information about the status code of the response sent to the user agent (since the proxy can modify it in certain circumstances) how do you suggest to pass that?

**Comment by vietj** [comment]
we can have handlers but they need to not use an async result, i.e `Handler<Event>`

File: src/main/java/io/vertx/httpproxy/HttpProxy.java [hunk 1]
```diff
@@ -80,6 +80,12 @@ default HttpProxy origin(int port, String host) {
   @Fluent
   HttpProxy originSelector(Function<HttpServerRequest, Future<SocketAddress>> selector);
 
+  @Fluent
+  HttpProxy onProxyRequestCompletion(Handler<Future<Void>> handler);
```
Comment (vietj) [inline src/main/java/io/vertx/httpproxy/HttpProxy.java:84]: that should be `Handler<Void>`

File: src/main/java/io/vertx/httpproxy/HttpProxy.java [hunk 2]
```diff
@@ -80,6 +80,12 @@ default HttpProxy origin(int port, String host) {
...
+
+  @Fluent
+  HttpProxy onProxyResponseCompletion(Handler<Future<Void>> handler);
```
Comment (vietj) [inline src/main/java/io/vertx/httpproxy/HttpProxy.java:87]: that should be `Handler<Void>`

File: src/main/java/io/vertx/httpproxy/HttpProxy.java (hunk 1 above)
Comment (dleen) [reply src/main/java/io/vertx/httpproxy/HttpProxy.java]: I went with the suggestion of `Handler<Event>`

**Review by jvican — APPROVED** [review]
LGTM

**Comment by jvican** [comment]
@vietj Any chance you can take another look and release?

**Comment by vietj** [comment]
@dleen I'm wondering if we should simply not instead of having event, expose the internal interceptor like API instead

**Comment by dleen** [comment]
@vietj Can you be more explicit about which internal interceptor API you are considering? Are you talking about the `ProxyRequest` and `ProxyResponse` interfaces? If so you might have some guidance. I added the `ProxyEvent` interface for the following situations:

1. In two places we are sending the request with a different response code than that of the proxied request. The client doesn't get to see that the status code has changed from `200` to `501` or `304`. What are your thoughts here?
```
if (chunked == null) {
  // response.request().release(); // Is it needed ???
  end(response.request(), 501);
  onHandleProxyResponse(ProxyEvent.fromResponse(response).withStatusCode(501)); // should use END future here
  return;
}
```

```
if (resource.lastModified.getTime() <= ifModifiedSince.getTime()) {
  outboundRequest.response().setStatusCode(304).end();
  onHandleProxyRequest(ProxyEvent.fromRequest(proxyRequest).withStatusCode(304));
  return true;
}
```


2. In other situations when the proxy request has failed the caller isn't able to know any details about the failed request:
```
if (ar.succeeded()) {
  // Allow consumers to handle the event first
  onHandleProxyRequest(ProxyEvent.fromRequest(request));
  sendProxyResponse(ar.result());
} else {
  onHandleProxyRequest(ProxyEvent.failedRequest(request));
}
```
I had considered passing `null` or something like that to indicate failure but it's not really useful for the consumer. What request failed?

3. I was worried about exposing `ProxyRequest`, `ProxyResponse` as they have mutable methods such as `release`, `send`, `setStatusCode` which it seems like the consumer should never be able to access and `send/release` probably have strange behavior if they are called after the request has already been sent or released.

What do you think is a good solution in light of the above concerns?

**Comment by dleen** [comment]
@vietj just a friendly ping 😄

**Comment by vietj** [comment]
I'm thinking about the current internal `ProxyContext` interface, would exposing that allow to do the same ?

**Comment by dleen** [comment]
@vietj I don't think so, the problem isn't with the methods of the `ProxyContext` interface i.e.
```
    void handleProxyRequest(ProxyRequest request);
    void handleProxyResponse(ProxyResponse response);
```
The issue is with the types of the parameters. `ProxyEvent` is introduced to overcome the issues with `ProxyRequest/ProxyResponse` mentioned in my earlier comment https://github.com/eclipse-vertx/vertx-http-proxy/pull/30#issuecomment-982938146.

**Comment by dleen** [comment]
@vietj Any further thoughts?

**Comment by jvican** [comment]
@dleen What you say above makes sense :+1:

@vietj Would be cool to get this work through the finish line, mind taking another look 😄

**Comment by vietj** [comment]
@dleen here is a PR with various enhancements https://github.com/eclipse-vertx/vertx-http-proxy/pull/32

among them there is now a filter API that should allow to implement the use case of #29 without extra API, do you mind having a look ?

**Comment by vietj** [comment]
shall be achieved byh https://github.com/eclipse-vertx/vertx-http-proxy/pull/32

**Pull request closed without merging**